rules/DeadCode/Rector/FunctionLike/TooWideReturnTypeRector.php

```diff
@@ -0,0 +1,243 @@
+
+        if ($this->shouldSkipNode($node, $scope)) {
+            return null;
+        }
+
+        $returnStatements = $this->betterNodeFinder->findInstanceOf($node, Return_::class);
+        $isAlwaysTerminating = $node instanceof ArrowFunction
+            || $this->terminatedNodeAnalyzer->isAlwaysTerminating($node);
+
+        if ($returnStatements === [] && ! $node instanceof ArrowFunction) {
+            $node->returnType = $isAlwaysTerminating
+                ? new Identifier('never')
+                : new Identifier('void');
+
+            return $node;
+        }
+
+        $returnType = $node->returnType;
+
+        if (! $returnType instanceof UnionType) {
+            return null;
+        }
+
+        $actualReturnTypes = $this->collectActualReturnTypes($node, $returnStatements, $isAlwaysTerminating);
+        $newReturnType = $this->narrowUnionReturnType($returnType, $actualReturnTypes);
+
+        if ($newReturnType === null) {
+            return null;
+        }
+
+        $node->returnType = $newReturnType;
+
+        return $node;
+    }
+
+    private function shouldSkipNode(ClassMethod|Function_|Closure|ArrowFunction $node, Scope $scope): bool
+    {
+        if ($this->hasYield($node)) {
+            return true;
+        }
+
+        if (! $node instanceof ClassMethod) {
+            return false;
+        }
+
+        if ($node->isPrivate() || $node->isFinal()) {
+            return false;
+        }
+
+        if ($node->isAbstract()) {
+            return true;
+        }
+
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+
+        if (! $classReflection instanceof ClassReflection) {
+            return true;
+        }
+
+        return ! $classReflection->isFinal();
+    }
+
+    /**
+     * @param Return_[] $returnStatements
+     * @return Type[]
+     */
+    private function collectActualReturnTypes(
+        ClassMethod|Function_|Closure|ArrowFunction $node,
+        array $returnStatements,
+        bool $isAlwaysTerminating,
+    ): array {
+        if ($node instanceof ArrowFunction) {
+            return [$this->getType($node->expr)];
+        }
+
+        $returnTypes = [];
+        foreach ($returnStatements as $returnStatement) {
+            if ($returnStatement->expr === null) {
+                $returnTypes[] = new NullType();
+                continue;
+            }
+
+            $returnTypes[] = $this->getType($returnStatement->expr);
+        }
+
+        if (! $isAlwaysTerminating) {
+            $returnTypes[] = new NullType();
+        }
+
+        return $returnTypes;
+    }
+
+    /**
+     * @param Type[] $actualReturnTypes
+     */
+    private function narrowUnionReturnType(
+        UnionType $unionType,
+        array $actualReturnTypes
+    ): ComplexType|Identifier|Name|null {
+        $types = $unionType->types;
+        $usedTypes = [];
+
+        foreach ($types as $type) {
+            $declaredType = $this->getType($type);
```

Comment:
use native type when possible:

```suggestion
            $declaredType = $type instanceof \PhpParser\Node\Expr 
                 ? $this->nodeTypeResolver->getNativeType($type)
                 : $this->getType($type);
```